src/include/index/ivf_pq_index.h

```diff
@@ -774,34 +763,46 @@ class ivf_pq_index {
         training_set_ids.end(),
         feature_vectors_.ids());
 
-    auto num_unique_labels = ::num_vectors(flat_ivf_centroids_);
```

Comment:
Conceptually, I would prefer to move `train_pq, train_ivf` in `train` and have add only perform the pq vector encoding and vector shuffling part. I think this will be a better approach to move to a potential OOC ingestion implementation

Reply:
Sure, seems reasonable, though I'd prefer to leave for now and improve along with out-of-core work if that's okay with you.

Reply:
sgtm